mayastor/tests/rebuild.rs

```diff
@@ -0,0 +1,1051 @@
+    assert_eq!(
+        get_nexus_state(nexus_hdl, NEXUS_UUID).await.unwrap(),
+        NexusState::NexusDegraded as i32
+    );
+
+    // Remove the child that is being rebuilt.
+    remove_child(nexus_hdl, child).await;
+    // Give a little time for the rebuild to fail.
+    std::thread::sleep(Duration::from_secs(1));
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+    // Nexus must be online because it has a single healthy child.
+    assert_eq!(
+        get_nexus_state(nexus_hdl, NEXUS_UUID).await.unwrap(),
+        NexusState::NexusOnline as i32
+    );
+}
+
+/// Test faulting the source of a rebuild.
+#[tokio::test]
+async fn rebuild_fault_src() {
+    let test = start_infrastructure("rebuild_fault_src").await;
+    let (mut ms1, mut ms2, ms3) = setup_test(&test, 1).await;
+    let nexus_hdl = &mut ms1;
+    let child = &get_share_uri(&ms3);
+
+    // Check a rebuild is started for the added child.
+    add_child(nexus_hdl, child, true).await;
+    assert!(wait_for_rebuild_state(
+        nexus_hdl,
+        child,
+        "running",
+        Duration::from_millis(500),
+    )
+    .await
+    .unwrap());
+
+    // Fault the rebuild source by unsharing the bdev.
+    bdev_unshare(&mut ms2).await;
+
+    // The rebuild failed so the destination should be faulted.
+    assert!(
+        wait_for_child_state(
+            nexus_hdl,
+            child,
+            ChildState::ChildFaulted,
+            Duration::from_millis(500),
+        )
+        .await
+    );
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+}
+
+/// Test faulting the destination of a rebuild.
+#[tokio::test]
+async fn rebuild_fault_dst() {
+    let test = start_infrastructure("rebuild_fault_dst").await;
+    let (mut ms1, _, mut ms3) = setup_test(&test, 1).await;
+    let nexus_hdl = &mut ms1;
+    let child = &get_share_uri(&ms3);
+
+    // Check a rebuild is started for the added child.
+    add_child(nexus_hdl, child, true).await;
+    assert!(wait_for_rebuild_state(
+        nexus_hdl,
+        child,
+        "running",
+        Duration::from_millis(500),
+    )
+    .await
+    .unwrap());
+
+    // Fault the rebuild destination by unsharing the bdev.
+    bdev_unshare(&mut ms3).await;
+
+    // Check the state of the destination child.
+    // Give a sufficiently high timeout time as unsharing an NVMf bdev can take
+    // some time to propagate up as an error from the rebuild job.
+    assert!(
+        wait_for_child_state(
+            nexus_hdl,
+            child,
+            ChildState::ChildFaulted,
+            Duration::from_secs(20),
+        )
+        .await
+    );
+    assert_eq!(
+        get_nexus_state(nexus_hdl, NEXUS_UUID).await.unwrap(),
+        NexusState::NexusDegraded as i32
+    );
+    assert_eq!(get_num_rebuilds(nexus_hdl).await, 0);
+}
+
+/// Test rebuild with different sizes of source and destination children.
+#[tokio::test]
+async fn rebuild_sizes() {
+    struct TestCase {
+        child1_size: Option<u64>,
```

Comment:
Why are these options? You can do bdev_create_and_share(ms2, **Some(**test.child1_size**)**...

Reply:
Good point - changed it.